<a href="https://colab.research.google.com/github/SelmaDM/Pyterrier/blob/master/Pyterrier_Ranking.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## PyTerrier - Learning to Rank

Ce TP à comme objetif la construction des pipelines dans [PyTerrier] (https://github.com/terrier-org/pyterrier). Toutes les expériences sont menées en utilisant le [corpus CORD19](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7251955/) et la [collection de tests TREC Covid](https://ir.nist.gov/covidSubmit/). Finalement, vous serez amené à refaire le processus avec une collection de questions-réponses.

Ce TP est divisé en deux. En particulier, dans la permière partie, vous allez expérimenter pour : 
 - comprendre le modèle de données PyTerrier ;
 - comprendre la notion générale de PyTerrier et les opérateurs permettant de combiner plusieurs recherches ;
 - entreprendre des expériences sur la collection de tests TREC Covid.

Dans la deuxième partie, vous ferez des expériences pour :
 - construire et apprendre des pipelines pour le Learning to Rank ;
 - évaluer et analyser des pipelines pour le Learning to Rank.


# Première partie

# Installation


In [49]:
!pip install -q --upgrade fastrank lightgbm==3.1.1
!pip install -q python-terrier

In [50]:
import pyterrier as pt
if not pt.started():
  pt.init()
cord19 = pt.datasets.get_dataset('irds:cord19/trec-covid')


# Indexation


In [51]:
!rm -rf ./terrier_cord19/*

In [52]:
import os

pt_index_path = './terrier_cord19'

if not os.path.exists(pt_index_path + "/data.properties"):
    # create the index, using the IterDictIndexer indexer 
    indexer = pt.index.IterDictIndexer(pt_index_path, blocks=True)

    # we give the dataset get_corpus_iter() directly to the indexer
    # while specifying the fields to index and the metadata to record
    index_ref = indexer.index(cord19.get_corpus_iter(), 
                              fields=('abstract',), 
                              meta=('docno',))

else:
    # if you already have the index, use it.
    index_ref = pt.IndexRef.of(pt_index_path + "/data.properties")

cord19/trec-covid documents:   0%|          | 0/192509 [00:00<?, ?it/s]

13:31:38.883 [ForkJoinPool-3-worker-3] WARN org.terrier.structures.indexing.Indexer - Adding an empty document to the index (6iu1dtyl) - further warnings are suppressed


<ipython-input-52-59cedea60614>:11: DeprecationWarning: specifying meta and meta_lengths in IterDictIndexer.index() is deprecated, use constructor instead
  index_ref = indexer.index(cord19.get_corpus_iter(),


13:33:17.619 [ForkJoinPool-3-worker-3] ERROR org.terrier.structures.indexing.Indexer - Could not finish MetaIndexBuilder: 
java.io.IOException: Key 8lqzfj2e is not unique: 37597,11755
For MetaIndex, to suppress, set metaindex.compressed.reverse.allow.duplicates=true
	at org.terrier.structures.collections.FSOrderedMapFile$MultiFSOMapWriter.mergeTwo(FSOrderedMapFile.java:1374)
	at org.terrier.structures.collections.FSOrderedMapFile$MultiFSOMapWriter.close(FSOrderedMapFile.java:1308)
	at org.terrier.structures.indexing.BaseMetaIndexBuilder.close(BaseMetaIndexBuilder.java:321)
	at org.terrier.structures.indexing.classical.BasicIndexer.indexDocuments(BasicIndexer.java:270)
	at org.terrier.structures.indexing.classical.BasicIndexer.createDirectIndex(BasicIndexer.java:388)
	at org.terrier.structures.indexing.Indexer.index(Indexer.java:377)
	at org.terrier.python.ParallelIndexer$3.apply(ParallelIndexer.java:131)
	at org.terrier.python.ParallelIndexer$3.apply(ParallelIndexer.java:120)
	at java.

## Operateurs

Vous aurez remarqué que BatchRetrieve possède une méthode `transform()` qui prend en entrée un dataframe, et retourne un autre dataframe, qui est en quelque sorte une *transformation* du dataframe précédent (par exemple, un résultat de la recherche). En fait, `BatchRetrieve` n'est qu'un des nombreux objets similaires dans PyTerrier, que nous appelons [transformers](https://pyterrier.readthedocs.io/en/latest/transformer.html) (représentés par la classe `TransformerBase`).
Par exemple, on peut sauvegarder explicitement dans la varible ```tfidf``` l'application du modèle tfidf


In [53]:
tfidf = pt.BatchRetrieve(index_ref, wmodel="TF_IDF")

La capacité intéressante de tous les transformateurs est qu'ils peuvent être combinés à l'aide d'opérateurs Python (c'est ce qu'on appelle la *surcharge d'opérateurs*).

Concrètement, imaginez que vous voulez enchaîner des transformateurs ensemble - par exemple, trier des documents d'abord par tf puis retrier les documents par tfidf. Nous pouvons le faire en utilisant l'opérateur `>>` - nous appelons cela *composition*.

Il existe un certain nombre d'opérateurs PyTerrier - il y a plus d'exemples dans la [documentation PyTerrier sur les opérateurs](https://pyterrier.readthedocs.io/en/latest/operators.html)




## Q1 - Construction de pipelines

Créez un *ranker* (une classe que permet de chercher sur l'indèxe) qui effectue les opérations suivantes :
 - obtenir les 10 documents les mieux notés par tf (`wmodel="Tf"`)
 - obtenir les 10 documents les mieux notés par tfidf (`wmodel="TF_IDF"`)
 - ne remet en ordre que les documents trouvés dans l'un ou l'autre des paramètres de recherche précédents en utilisant BM25.

en utilisant des opérateurs PyTerrier combinant différentes instances de BatchRetrieve.

## Q2 - \# documents
Combien de documents sont récupérés par ce pipeline complet pour la requête `"chemical"` ?
> Hint: Si vous obtenez la bonne solution, le document avec le docno `"8hykq71k"` devrait avoir un score proche de $12.646089$ pour la requête `"chemical"`.

Conseils :
 - choisissez soigneusement vos [opérateurs PyTerrier](https://pyterrier.readthedocs.io/en/latest/operators.html)
 - vous ne devriez pas avoir à effectuer d'opérations sur les dataframes.

In [54]:
# réponse Q1
bm25 = pt.BatchRetrieve(index_ref, wmodel="BM25")%10
tf = pt.BatchRetrieve(index_ref, wmodel="Tf")%10
tf_idf = pt.BatchRetrieve(index_ref, wmodel="TF_IDF")%10
pipeline = bm25 >> (tf ** tf_idf)

In [55]:
# réponse Q2
 #Nombre de documents recupérés : 1006 

In [56]:
bm25 = pt.BatchRetrieve(index_ref, wmodel="BM25")
tf = pt.BatchRetrieve(index_ref, wmodel="Tf")
tf_idf = pt.BatchRetrieve(index_ref, wmodel="TF_IDF")
pipeline = bm25 >> (tf ** tf_idf)

pipeline.search("chemical")

qid   docid     docno  rank      score     query  \
0      1   37771  jn5qi1jb     0  12.426309  chemical   
1      1   15671  8hykq71k     1  12.413269  chemical   
2      1  134305  0smev8vt     2  12.292890  chemical   
3      1  142104  77c9ohxj     3  12.226076  chemical   
4      1   87642  ck6clsty     4  12.155804  chemical   
...   ..     ...       ...   ...        ...       ...   
1001   1  192405  rh9h2rtz   995   6.065540  chemical   
1002   1    4713  gve33v4t   996   6.040683  chemical   
1003   1    9767  pcu5sy6t   997   6.040683  chemical   
1004   1   35909  xic0qqio   998   6.040683  chemical   
1005   1   45510  i5z4yq4p   999   6.040683  chemical   

                       features  
0       [9.0, 6.80992310099522]  
1      [4.0, 6.802776877224232]  
2      [8.0, 6.736806385246307]  
3      [7.0, 6.700190253577284]  
4      [5.0, 6.661679714989656]  
...                         ...  
1001   [1.0, 3.324065127180903]  
1002  [1.0, 3.3104430955870283]  
1003  [1.0, 3.3104430955870283]  
1004  [1.0, 3.3104430955870283]  
1005  [1.0, 3.3104430955870283]  

[1006 rows x 7 columns]

# Évaluation dans un pipeline
À difference du TP3 où nous avons fait nous même l'évaluation, dans ce TP nous allons utiliser le module d'évaluation intégré pour les pipelines en pyterrier. Donc, pour faire des expériences (évaluer un modèle) sur pyterrier, on peut utiliser la classe `Experiment`. Voici le code qui évalue la performance de tfidf pour la collection cord19.  

In [57]:
# Télécharger la collection

topics = cord19.get_topics(variant='description')
qrels = cord19.get_qrels()


In [58]:
len(topics), len(qrels)

(50, 69318)

In [59]:
# Voici le code pour évaluer avec Experiment
from pyterrier.measures import *
pt.Experiment(
  #le pipeline
  [tfidf],
  topics,
  qrels,
  eval_metrics=[MAP, nDCG, nDCG@10],
  # on utilise TFIDF pour les tests statistiques
  baseline=0,
  names=["TFIDF"]
)



name        AP      nDCG   nDCG@10  AP +  AP - AP p-value nDCG + nDCG -  \
0  TFIDF  0.188578  0.400915  0.638799  None  None       None   None   None   

  nDCG p-value nDCG@10 + nDCG@10 - nDCG@10 p-value  
0         None      None      None            None

# Deuxième Partie  – Learning to Rank

Dans cette partie du TP, vous ferez la construction, l'apprentissage,  l'évaluation et l'analyse des pipelines d'apprentissage du classement (Learning to Rank).


Tout d'abord, répartissons les annotations (rankings) en ensembles d'entraînement, de validation et de test. TREC Covid ne comporte que 50 annotations, ce qui n'est pas beaucoup pour l'appprentissage. Nous allons répartir 30 annotations pour l'entraînement, 5 pour la validation et 15 pour l'évaluation. Nous examinerons également les differences statistiques, même si elle est petite, pour 15 annotations.

Nous n'allons classer que les 10 premiers documents pour chaque requête - nous espérons que l'apprentissage du classement pourra nous aider à retrier les 10 premiers documents pour qu'ils soient plus efficaces.


In [60]:
RANK_CUTOFF = 10
SEED=42

from sklearn.model_selection import train_test_split

tr_va_topics, test_topics = train_test_split(topics, test_size=15, random_state=SEED)
train_topics, valid_topics =  train_test_split(tr_va_topics, test_size=5, random_state=SEED)


test_qrels = qrels # seulement les annotations des topics en réponse sont utilisés, donc pas de problème si on utilise tout
train_qrels = qrels
valid_qrels = qrels

## Ensemble de caractéristiques

Définissons notre ensemble de caractéristiques.  Nous allons avoir un total de 6 caractéristiques :

1. le score abstrait de TFIDF ;
2. le résumé contient-il "coronavirus covid", noté par TFIDF ;
3. le score TFIDF sur le titre (même si nous ne l'avons pas indexé plus tôt !);
4. l'article a-t-il été publié en 2020 ? Ici, on a l'hypothèse que les articles récents étaient plus utiles pour cette tâche ;
5. L'article a-t-il un DOI, c'est-à-dire s'agit-il d'une publication officielle ?
6. le score de correspondance des coordonnées pour la requête - c'est-à-dire combien de termes de la requête apparaissent dans le résumé.

Plusieurs de ces caractéristiques nécessitent des métadonnées supplémentaires `["title", "date", "doi"]`. Heureusement, le jeu de données TREC Covid nous permet d'obtenir plus de métadonnées après l'indexation. Nous utilisons `pt.text.get_text(cord19, ["title", "date", "doi"])` pour récupérer ces colonnes de métadonnées supplémentaires.



In [61]:
ltr_feats1 = (tfidf % RANK_CUTOFF) >> pt.text.get_text(cord19, ["title", "date", "doi"]) >> (
    pt.transformer.IdentityTransformer()
    ** # score of text for query 'coronavirus covid'
    (pt.apply.query(lambda row: 'coronavirus covid') >> tfidf)
    ** # score of title (not originally indexed)
    (pt.text.scorer(body_attr="title", takes='docs', wmodel='TF_IDF') ) 
    ** # date 2020
    (pt.apply.doc_score(lambda row: int("2020" in row["date"])))
    ** # has doi
    (pt.apply.doc_score(lambda row: int( row["doi"] is not None and len(row["doi"]) > 0) ))
    ** # abstract coordinate match
    pt.BatchRetrieve(index_ref, wmodel="CoordinateMatch")
)

# for reference, lets record the feature names here too
fnames=["TFIDF", 'coronavirus covid', 'title', "2020", "hasDoi", "CoordinateMatch"]

Voyons le résultat pour une requête particulière. Nous pouvons voir que nous avons maintenant des colonnes supplémentaires de métadonnées de document `["title", "date", "doi"]`, ainsi que les colonnes `"features"` très importantes. Cela fait que le dataframe a le type $R_f$. En effet, c'est cette colonne que nous utilisons pour l'apprentissage.

In [62]:
ltr_feats1.search("Movie")

13:33:41.592 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


qid   docid     docno  rank      score  query  \
0   1   23347  qiwq0pe5     0  12.970648  Movie   
1   1   23343  vmetwotq     1  12.923459  Movie   
2   1   78848  mmq44kwx     2  12.207116  Movie   
3   1   70132  eynhsuz8     3  11.372312  Movie   
4   1   24731  o7ckdng4     4  10.805086  Movie   
5   1  118935  25khbzk0     5   9.594602  Movie   
6   1   16001  3lhpdpiv     6   9.058977  Movie   
7   1   18722  gt3xayqp     7   8.579994  Movie   
8   1   86076  opbwnnai     8   8.242861  Movie   
9   1    5238  5z3pbbfb     9   8.042828  Movie   

                                               title        date  \
0  Sentiment Analysis on Movie Scripts and Review...  2020-05-06   
1  Improving Movie Recommendation Systems Filteri...  2020-05-04   
2  Smoking in top-grossing movies--United States,...        2011   
3    The Post: A token woman leader's transformation        2020   
4   Movies Emotional Analysis Using Textual Contents  2020-05-26   
5  CinemaGazer: a System for Watching Video at Ve...  2011-10-04   
6  The Aliens in Us and the Aliens Out There: Sci...  2013-11-17   
7  The Unfairness of Popularity Bias in Music Rec...  2020-03-24   
8  Preparing for an influenza pandemic: mental he...        2009   
9  Characteristics of airborne Staphylococcus aur...  2014-05-07   

                            doi  \
0  10.1007/978-3-030-49161-1_36   
1  10.1007/978-3-030-49190-1_17   
2                                 
3            10.1002/hrdq.21391   
4  10.1007/978-3-030-51310-8_19   
5                                 
6   10.1007/978-1-4614-7175-2_2   
7   10.1007/978-3-030-45442-5_5   
8                                 
9     10.1007/s10453-014-9342-6   

                                            features  
0  [12.97064764826613, 0.0, 0.7458647062217206, 1...  
1  [12.923458775835545, 0.0, 1.0350775514913675, ...  
2  [12.207115891520253, 0.0, 0.8744620693633965, ...  
3  [11.372311997669033, 1.499168191418449, 0.0, 1...  
4  [10.80508574921903, 0.0, 0.9277829272514087, 1...  
5       [9.594602481582845, 0.0, 0.0, 0.0, 0.0, 1.0]  
6  [9.058977465455175, 0.0, 0.9880285718781234, 0...  
7       [8.579993656638317, 0.0, 0.0, 1.0, 1.0, 1.0]  
8       [8.242860905714242, 0.0, 0.0, 0.0, 0.0, 1.0]  
9       [8.042828188746888, 0.0, 0.0, 0.0, 1.0, 1.0]

Nous pouvons également examiner les valeurs brutes des caractéristiques (dans ce cas, pour le premier document classé). Notez que le BM25 dans la colonne "score" ci-dessus est également la première valeur dans le tableau des caractéristiques (proche de 12), parce que nous avons utilisé un transformateur d'identité.

In [63]:
ltr_feats1.search("Movie").iloc[0]["features"]

13:33:41.995 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


array([12.97064765,  0.        ,  0.74586471,  1.        ,  1.        ,
        1.        ])

## Analysis

Ici nous analisons les performances de chaque caractéristique indépendamment. Pour cela, nous composons le pipeline de caractéristiques (`ltr_feats1`) avec `pt.ltr.feature_to_score(i)` pour un certain nombre de caractéristiques $i$.

In [64]:
pt.Experiment(
    [ltr_feats1 >> pt.ltr.feature_to_score(i) for i in range(len(fnames))],
    test_topics,
    test_qrels,  
    names=fnames,
    eval_metrics=["map", "ndcg", "ndcg_cut_10", "num_rel_ret"])

13:33:45.340 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


13:33:51.151 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


13:33:53.677 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


13:33:56.113 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


13:33:58.541 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


13:34:02.439 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


name       map      ndcg  ndcg_cut_10  num_rel_ret
0              TFIDF  0.010519  0.047832     0.589368         96.0
1  coronavirus covid  0.010816  0.047858     0.584780         96.0
2              title  0.012200  0.054099     0.647125         96.0
3               2020  0.010942  0.048751     0.591473         96.0
4             hasDoi  0.010437  0.047810     0.575902         96.0
5    CoordinateMatch  0.010204  0.046617     0.570048         96.0

Il est intéressant de noter que la caractéristique "coronavirus covid" a obtenu un NDCG@10 de 0,5847. Il s'agit donc d'une baseline forte pour cette tâche.



## Learning 

Dans cette partie du TP, nous appliquons trois techniques différentes d'apprentissage du classement :

 - l'ascension des coordonnées à partir de FastRank, une technique linéaire par liste
 - les forêts aléatoires de `scikit-learn`, une technique d'arbre de régression par liste
 - LambdaMART de LightGBM, une technique d'arbre de régression par liste.

Dans chaque cas, nous prenons notre pipeline de caractéristiques, `ltr_feats1`, et nous le composons (`>>`) avec le modèle appris. Nous utilisons `pt.ltr.apply_learned_model()` qui sait comment gérer les différents apprenants.

Le pipeline complet est ensuite ajusté (appris) en utilisant `.fit()`, en spécifiant les annotations d'entraînement et les qrels. Il est important de noter que les étapes précédentes du pipeline (récupération et calcul des caractéristiques) sont appliquées aux annotations d'entraînement afin d'obtenir les résultats, qui sont ensuite transmis à la technique d'apprentissage par classement. LightGBM a un arrêt précoce activé, qui utilise un ensemble de annotations de validation - de la même manière, les annotations de validation sont transformés en résultats de validation.

Enfin, `%time` est la "commande magique" qui affiche le temps d'apprentissage pour chaque technique. L'apprentissage pour chaque technique prend < 30 secondes.

In [65]:
import fastrank

train_request = fastrank.TrainRequest.coordinate_ascent()

params = train_request.params
params.init_random = True
params.normalize = True
params.seed = 1234567

ca_pipe = ltr_feats1 >> pt.ltr.apply_learned_model(train_request, form='fastrank')

%time ca_pipe.fit(train_topics, train_qrels)

13:34:06.100 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 300 received 306, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


CPU times: user 30.9 s, sys: 110 ms, total: 31 s
Wall time: 19.7 s


In [66]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(n_estimators=400, verbose=1, random_state=42, n_jobs=2)

rf_pipe = ltr_feats1 >> pt.ltr.apply_learned_model(rf)


%time rf_pipe.fit(train_topics, train_qrels)

13:34:26.277 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 300 received 306, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "
[Parallel(n_jobs=2)]: Using backend ThreadingBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  46 tasks      | elapsed:    0.3s
[Parallel(n_jobs=2)]: Done 196 tasks      | elapsed:    1.2s


CPU times: user 9.06 s, sys: 126 ms, total: 9.19 s
Wall time: 7.42 s


[Parallel(n_jobs=2)]: Done 400 out of 400 | elapsed:    2.3s finished


In [67]:
import lightgbm as lgb

# this configures LightGBM as LambdaMART
lmart_l = lgb.LGBMRanker(
    task="train",
    silent=False,
    min_data_in_leaf=1,
    min_sum_hessian_in_leaf=1,
    max_bin=255,
    num_leaves=31,
    objective="lambdarank",
    metric="ndcg",
    ndcg_eval_at=[10],
    ndcg_at=[10],
    eval_at=[10],
    learning_rate= .1,
    importance_type="gain",
    num_iterations=100,
    early_stopping_rounds=5
)

lmart_x_pipe = ltr_feats1 >> pt.ltr.apply_learned_model(lmart_l, form="ltr", fit_kwargs={'eval_at':[10]})

%time lmart_x_pipe.fit(train_topics, train_qrels, valid_topics, valid_qrels)

13:34:33.217 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 300 received 306, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


13:34:35.053 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 50 received 54, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


[LightGBM] [Warning] early_stopping_round is set=5, early_stopping_rounds=5 will be ignored. Current value: early_stopping_round=5
[LightGBM] [Warning] min_sum_hessian_in_leaf is set=1, min_child_weight=0.001 will be ignored. Current value: min_sum_hessian_in_leaf=1
[LightGBM] [Warning] min_data_in_leaf is set=1, min_child_samples=20 will be ignored. Current value: min_data_in_leaf=1
[LightGBM] [Warning] min_sum_hessian_in_leaf is set=1, min_child_weight=0.001 will be ignored. Current value: min_sum_hessian_in_leaf=1
[LightGBM] [Warning] min_data_in_leaf is set=1, min_child_samples=20 will be ignored. Current value: min_data_in_leaf=1
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000929 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 268
[LightGBM] [Info] Number of data points in the train set: 24870, number of used features: 6
[Li

/usr/local/lib/python3.9/dist-packages/lightgbm/engine.py:151: UserWarning: Found `num_iterations` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))
/usr/local/lib/python3.9/dist-packages/lightgbm/engine.py:156: UserWarning: Found `early_stopping_rounds` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))


## Evaluation

Comparons maintenant nos pipelines de classement sur nos 15 annotations de teste versus la référence de base BM25. Dans tous les cas, nous ne classons que 10 résultats par requête, donc le MAP sera carrément plus faible. 

Nous allons rapporter le temps de réponse moyen (`"mrt"`) ainsi que les mesures MAP, NDCG et NDCG@10.

In [68]:
pt.Experiment(
    [tfidf % RANK_CUTOFF, ca_pipe, rf_pipe, lmart_x_pipe],
    test_topics,
    test_qrels, 
    names=["TFIDF",  "TFIDF + CA(6f)", "TFIDF + RF(6f)", "TFIDF + LMart(6f)"],
    baseline=0,
    eval_metrics=["map", "ndcg", "ndcg_cut_10", "mrt"])

13:34:38.684 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


13:34:42.011 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "
[Parallel(n_jobs=2)]: Using backend ThreadingBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  46 tasks      | elapsed:    0.0s
[Parallel(n_jobs=2)]: Done 196 tasks      | elapsed:    0.1s
[Parallel(n_jobs=2)]: Done 400 out of 400 | elapsed:    0.2s finished


13:34:44.640 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


name       map      ndcg  ndcg_cut_10         mrt  map +  \
0              TFIDF  0.010519  0.047832     0.589368   72.828908    NaN   
1     TFIDF + CA(6f)  0.011807  0.052092     0.630170  191.067751    9.0   
2     TFIDF + RF(6f)  0.011907  0.051449     0.627952  203.712403    8.0   
3  TFIDF + LMart(6f)  0.011828  0.052731     0.622488  151.677864    6.0   

   map -  map p-value  ndcg +  ndcg -  ndcg p-value  ndcg_cut_10 +  \
0    NaN          NaN     NaN     NaN           NaN            NaN   
1    1.0     0.052802     9.0     1.0      0.010754            9.0   
2    4.0     0.038993     9.0     3.0      0.090934            9.0   
3    5.0     0.217733     5.0     6.0      0.161192            5.0   

   ndcg_cut_10 -  ndcg_cut_10 p-value  
0            NaN                  NaN  
1            1.0             0.006948  
2            3.0             0.078531  
3            6.0             0.223187

Question - les trois modèles appris ont pu améliorer NDCG@10 sur TFIDF ? Analysez les résultats, en indicant s'il y a des amélorations statistiques.

**Réponse**
Oui, les 3 modèles ont tous amelioré le **ndcg_cut_10** qui est passé de "0.589368" **à** "0.6337"  / "0.6279" / "0.6336" selon le modèle 

## Tâche pratique - Concaténation

Notre modèle appris a un faible rappel car seulement 10 documents sont reclassés. Créons une petite fonction, `append_baseline()`, qui peut ajouter les résultats de la ligne de base BM25 à la sortie du modèle appris. Ceci est défini en utilisant les [opérateurs de transformation] (https://pyterrier.readthedocs.io/en/latest/operators.html) (`^` et `%`).

Comme exercice, appliquez `append_baseline()` à chacun des pipelines de modèles appris définis ci-dessus, et rapportez le MAP et le NDCG calculés sur les 1000 résultats classés. 

Lequel des modèles appris résulte en une amélioration significative de MAP et NDCG ?

**Réponse** : Modèle numéro 01 


In [69]:
def append_baseline(system, baseline, max_results = 1000):
    return (system ^ baseline) % max_results

In [70]:
#réponse

result1 = append_baseline(ca_pipe,bm25)
result2 = append_baseline(rf_pipe,bm25)
result3 = append_baseline(lmart_x_pipe,bm25)
pt.Experiment(
    [tfidf % RANK_CUTOFF, result1, result2, result3],
    test_topics,
    test_qrels, 
    names=["TFIDF",  "TFIDF + CA(6f)", "TFIDF + RF(6f)", "TFIDF + LMart(6f)"],
    baseline=0,
    eval_metrics=["map", "ndcg"])

13:34:48.204 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


13:34:52.356 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "
[Parallel(n_jobs=2)]: Using backend ThreadingBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  46 tasks      | elapsed:    0.0s
[Parallel(n_jobs=2)]: Done 196 tasks      | elapsed:    0.2s
[Parallel(n_jobs=2)]: Done 400 out of 400 | elapsed:    0.4s finished


13:34:57.141 [main] WARN org.terrier.querying.ApplyTermPipeline - The index has no termpipelines configuration, and no control configuration is found. Defaulting to global termpipelines configuration of 'Stopwords,PorterStemmer'. Set a termpipelines control to remove this warning.


/usr/local/lib/python3.9/dist-packages/pyterrier/ops.py:242: UserWarning: Got number of results different expected from <pyterrier.batchretrieve.TextScorer object at 0x7f5120e57e80>, expected 150 received 154, feature scores for any missing documents be 0, extraneous documents will be removed
  warn("Got number of results different expected from %s, expected %d received %d, feature scores for any "


name       map      ndcg  map +  map -  map p-value  ndcg +  \
0              TFIDF  0.010519  0.047832    NaN    NaN          NaN     NaN   
1     TFIDF + CA(6f)  0.141777  0.333775   14.0    1.0     0.000051    14.0   
2     TFIDF + RF(6f)  0.141749  0.332501   14.0    1.0     0.000055    14.0   
3  TFIDF + LMart(6f)  0.141651  0.333769   13.0    2.0     0.000056    13.0   

   ndcg -  ndcg p-value  
0     NaN           NaN  
1     1.0      0.000002  
2     1.0      0.000003  
3     2.0      0.000003







# Tâche pratique

Utilisez les modèles implémentés pour cord19 dans une tâche de question-réponse. Dans ce contexte, les requêtes sont de questions et les documents sont des documents qui pourraient contenir la réponse. Notez que vous devez refaire l'indexation ainsi que les autres étapes étudiées dans ce TP. Voici un exemple du dataset à utiliser :

```json
"question": "Why are big companies like Apple or Google not included in the Dow Jones Industrial Average (DJIA) index?",

"answers":{
  "290156": { 
    "text":" That is a pretty exclusive club and for the most part they are not interested in highly volatile companies like Apple and Google. Sure, IBM is part of the DJIA, but that is about as stalwart as you can get these days. The typical profile for a DJIA stock would be one that pays fairly predictable dividends, has been around since money was invented, and are not going anywhere unless the apocalypse really happens this year. In summary, DJIA is the boring reliable company index." ,
    "timestamp": "Sep 11 '12 at 0:53"}
 }

```
Vous pouvez le télécharger en utilisant les lines de code ci-dessous.

In [71]:
fiqa = {}
fiqa['train'] = pt.datasets.get_dataset('irds:beir/fiqa/train')
fiqa['valid'] = pt.datasets.get_dataset('irds:beir/fiqa/dev')
fiqa['test'] = pt.datasets.get_dataset('irds:beir/fiqa/test')

test_topics = fiqa['test'].get_topics(variant='text')
test_qrels = fiqa['test'].get_qrels()

train_topics = fiqa['train'].get_topics(variant='text')
train_qrels = fiqa['train'].get_qrels()

valid_topics = fiqa['valid'].get_topics(variant='text')
valid_qrels = fiqa['valid'].get_qrels()

In [72]:
#ltr_feats1
print(fiqa)
print(train_topics)
print(train_qrels)
test = pt.text.scorer()

{'train': IRDSDataset('beir/fiqa/train'), 'valid': IRDSDataset('beir/fiqa/dev'), 'test': IRDSDataset('beir/fiqa/test')}
        qid                                              query
0         0  what is considered a business expense on a bus...
1         4  business expense car insurance deductible for ...
2         5                     starting a new online business
3         6                business day and due date for bills
4         7  new business owner how do taxes work for the b...
...     ...                                                ...
5495  11092  is real physical money traded during online tr...
5496  11096                  pensions annuities and retirement
5497  11097  what does it mean to a life insurance policy h...
5498  11099  can capital loss in traditional ira and roth i...
5499  11104  selling a stock for gain to offset other stock...

[5500 rows x 2 columns]
         qid   docno  label iteration
0          0   18850      1         0
1          4  196463   

In [73]:
# Creation des fichiers d'indexation
!rm -rf ./pd_index_fiqa_train
!rm -rf ./pd_index_fiqa_test
!rm -rf ./pd_index_fiqa_valid

In [74]:
def indexingFun(data_list):
  pt_index_path_dict = {}
  for data in data_list:
    pt_index_path = './pd_index_fiqa' + '_' + data
    pt_index_path_dict[data] = pt_index_path

    if not os.path.exists(pt_index_path + "/data.properties"):
        # create the index, using the IterDictIndexer indexer 
        indexer = pt.index.IterDictIndexer(pt_index_path, blocks=True)

        # we give the dataset get_corpus_iter() directly to the indexer
        # while specifying the fields to index and the metadata to record
        index_ref = indexer.index(fiqa[data].get_corpus_iter())

    else:
        # if you already have the index, use it.
        index_ref = pt.IndexRef.of(pt_index_path + "/data.properties")
  return pt_index_path_dict

In [75]:
def tpPipeline(index_ref):
  tf = pt.BatchRetrieve(index_ref, wmodel="Tf")
  tfidf = pt.BatchRetrieve(index_ref, wmodel="TF_IDF")
  bm_25 = pt.BatchRetrieve(index_ref, wmodel='BM25')
  
  pipeline = ((tf %10) | (tfidf % 10)) >> bm_25

  return pipeline

In [76]:
#pipeline de caractéristiques 
def tpCharactPipeline(pipeline, data, index_ref, RANK_CUTOFF=10):
  return (pipeline % RANK_CUTOFF) >> pt.text.get_text(data, metadata=['doc_id', 'text'], by_query=True) >> (
      pt.transformer.IdentityTransformer()
      ** 
      pt.BatchRetrieve(index_ref, wmodel="CoordinateMatch")
      )

In [77]:
import fastrank


def fastRankModel(charact_pipeline, train_topics, train_qrels, validation_topics, validation_qrels):
  train_request = fastrank.TrainRequest.coordinate_ascent()

  params = train_request.params
  params.init_random = True
  params.normalize = True
  params.seed = 1234567

  ca_pipe = charact_pipeline >> pt.ltr.apply_learned_model(train_request, form='fastrank')

  %time ca_pipe.fit(train_topics, train_qrels, validation_topics, validation_qrels)

  return ca_pipe

In [78]:
from sklearn.ensemble import RandomForestRegressor


def randomForestModel(charact_pipeline, train_topics, train_qrels, validation_topics, validation_qrels):
  rf = RandomForestRegressor(n_estimators=400, verbose=1, random_state=42, n_jobs=2)

  rf_pipe = charact_pipeline >> pt.ltr.apply_learned_model(rf)
  %time rf_pipe.fit(train_topics, train_qrels, validation_topics, validation_qrels)
  return rf_pipe

In [79]:
import lightgbm as lgb


def lambdaRankModel(charact_pipeline, train_topics, train_qrels, validation_topics, validation_qrels):
  # this configures LightGBM as LambdaMART
  lmart_l = lgb.LGBMRanker(
      task="train",
      silent=False,
      min_data_in_leaf=1,
      min_sum_hessian_in_leaf=1,
      max_bin=255,
      num_leaves=31,
      objective="lambdarank",
      metric="ndcg",
      ndcg_eval_at=[10],
      ndcg_at=[10],
      eval_at=[10],
      learning_rate= .1,
      importance_type="gain",
      num_iterations=100,
      early_stopping_rounds=5
  )

  lmart_x_pipe = charact_pipeline >> pt.ltr.apply_learned_model(lmart_l, form="ltr", fit_kwargs={'eval_at':[10]})
  %time lmart_x_pipe.fit(train_topics, train_qrels, validation_topics, validation_qrels)

  return lmart_x_pipe

In [80]:
from pyterrier.measures import *

def evalPipeline(pipeline, test_topics, test_qrels, name):
  return pt.Experiment(
      [pipeline],
      test_topics,
      test_qrels,
      eval_metrics=[MAP, nDCG, nDCG@10],

      baseline=0,
      names=name
      )

In [81]:
def evalCharctPipeline(charact_pipeline, feature_names, test_topics, test_qrels):
  return pt.Experiment(
      [charact_pipeline >> pt.ltr.feature_to_score(i) for i in range(len(feature_names))],
      test_topics,
      test_qrels,  
      names=feature_names,
      eval_metrics=["map", "ndcg", "ndcg_cut_10", "num_rel_ret"])

In [82]:
def evalModel(pipeline, model_pipelines, test_topics, test_qrels, names, RANK_CUTOFF=10):
  return pt.Experiment(
      [pipeline % RANK_CUTOFF] + model_pipelines,
      test_topics,
      test_qrels, 
      names=names,
      baseline=0,
      eval_metrics=["map", "ndcg", "ndcg_cut_10", "mrt"])

In [83]:
def evalModelWithAppend(pipeline, model_pipelines, test_topics, test_qrels, names):
  return pt.Experiment(
      [pipeline] + [append_baseline(x, pipeline) for x in model_pipelines],
      test_topics,
      test_qrels, 
      names=names,
      baseline=0,
      eval_metrics=["map", "ndcg", "ndcg_cut_10", "mrt"])

In [ ]:
data_list = ['train', 'test', 'valid']

names = ['Pipeline', 'FastR', 'RFR', 'Lambda']
feature_names = ['CoordinateMatch']

# Indexation
pt_index_path_dict = indexingFun(data_list)

# Creation Pipeline
pipeline = tpPipeline(pt.IndexRef.of(pt_index_path_dict['train'] + "/data.properties"))
charact_pipeline = tpCharactPipeline(pipeline, fiqa['train'], pt_index_path_dict['train'])
model_pipelines = [fastRankModel(charact_pipeline, train_topics, train_qrels, valid_topics, valid_qrels),
                   randomForestModel(charact_pipeline, train_topics, train_qrels, valid_topics, valid_qrels),
                   lambdaRankModel(charact_pipeline, train_topics, train_qrels, valid_topics, valid_qrels)]

# Evaluations
evaluations_dict = {}
print("Evaluation : Pipeline")
evaluations_dict['Pipeline'] = evalPipeline(pipeline, test_topics, test_qrels, ['Pipeline'])

print("Evaluation : Characteristics Pipeline")
evaluations_dict['Charact_Pipeline'] = evalCharctPipeline(charact_pipeline, feature_names, test_topics, test_qrels)

print("Evaluation : Models")
evaluations_dict['Models'] = evalModel(pipeline, model_pipelines, test_topics, test_qrels, names)

print("Evaluation : Models with append_baseline")
evaluations_dict['Madels_with_Append'] = evalModelWithAppend(pipeline, model_pipelines, test_topics, test_qrels, names)




beir/fiqa/train documents:   0%|          | 0/57638 [00:00<?, ?it/s]

13:35:00.701 [ForkJoinPool-4-worker-3] WARN org.terrier.structures.indexing.Indexer - Adding an empty document to the index (7915) - further warnings are suppressed
13:35:23.007 [ForkJoinPool-4-worker-3] WARN org.terrier.structures.indexing.Indexer - Indexed 39 empty documents


beir/fiqa/test documents:   0%|          | 0/57638 [00:00<?, ?it/s]

13:35:23.556 [ForkJoinPool-5-worker-3] WARN org.terrier.structures.indexing.Indexer - Adding an empty document to the index (7915) - further warnings are suppressed
13:35:44.066 [ForkJoinPool-5-worker-3] WARN org.terrier.structures.indexing.Indexer - Indexed 39 empty documents


beir/fiqa/dev documents:   0%|          | 0/57638 [00:00<?, ?it/s]

13:35:44.456 [ForkJoinPool-6-worker-3] WARN org.terrier.structures.indexing.Indexer - Adding an empty document to the index (7915) - further warnings are suppressed
13:36:06.011 [ForkJoinPool-6-worker-3] WARN org.terrier.structures.indexing.Indexer - Indexed 39 empty documents
CPU times: user 9min 15s, sys: 10.3 s, total: 9min 26s
Wall time: 9min 21s


[Parallel(n_jobs=2)]: Using backend ThreadingBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  46 tasks      | elapsed:    4.7s
[Parallel(n_jobs=2)]: Done 196 tasks      | elapsed:   19.8s
[Parallel(n_jobs=2)]: Done 400 out of 400 | elapsed:   40.4s finished


CPU times: user 9min 59s, sys: 9.02 s, total: 10min 8s
Wall time: 9min 41s


In [ ]:
evaluations_dict['Models']

In [ ]:
evaluations_dict["Madels_with_Append"]